sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java

```diff
@@ -0,0 +1,197 @@
+    broker = new BrokerService();
+    broker.setUseJmx(false);
+    broker.setPersistenceAdapter(new MemoryPersistenceAdapter());
+    broker.addConnector(new URI("mqtt://localhost:" + port));
+    broker.start();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<byte[]> output = pipeline.apply(
+        MqttIO.read()
+            .withConnectionConfiguration(
+                MqttIO.ConnectionConfiguration.create(
+                    "tcp://localhost:" + port,
+                    "READ_TOPIC"))
+          .withMaxNumRecords(10));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<byte[]>globally()))
+        .isEqualTo(10L);
+    PAssert.that(output).satisfies(new SerializableFunction<Iterable<byte[]>, Void>() {
+      @Override
+      public Void apply(Iterable<byte[]> input) {
+        for (byte[] element : input) {
+          String inputString = new String(element);
+          Assert.assertTrue(inputString.startsWith("This is test "));
+          int count = Integer.parseInt(inputString.substring("This is test ".length()));
+          Assert.assertTrue(count < 10);
+          Assert.assertTrue(count >= 0);
+        }
+        return null;
+      }
+    });
+
+    // produce messages on the broker in another thread
+    // This thread prevents to block the pipeline waiting for new messages
+    Thread thread = new Thread() {
+      public void run() {
+        try {
+          // gives time to the pipeline to start
+          Thread.sleep(2000);
+        } catch (Exception e) {
+          // nothing to do
+        }
+        try {
+          MqttClient client = new MqttClient("tcp://localhost:" + port,
+              MqttClient.generateClientId());
+          client.connect();
+          for (int i = 0; i < 10; i++) {
+            MqttMessage message = new MqttMessage();
+            message.setQos(0);
+            message.setPayload(("This is test " + i).getBytes());
+            client.publish("READ_TOPIC", message);
+          }
+          client.disconnect();
+          client.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+    };
+    thread.start();
+
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    List<MqttMessage> messages = new ArrayList<>();
+    MqttClient client = receive(messages);
+
+    Pipeline pipeline = TestPipeline.create();
+
+    ArrayList<byte[]> data = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      data.add("Test".getBytes());
+    }
+    pipeline.apply(Create.of(data))
+        .apply(MqttIO.write()
+            .withConnectionConfiguration(
+                MqttIO.ConnectionConfiguration.create(
+                    "tcp://localhost:" + port,
+                    "WRITE_TOPIC"))
+        .withQoS(1));
+    pipeline.run();
+
+    Assert.assertEquals(100, messages.size());
```

Comment:
Not with QoS 1: with this QoS, the publisher (so the pipeline), will wait the ACK from the broker (on the `publish()` method). The broker will send the ack to the pipeline only when the subscriber has received the messages (so added in the messages list).